### PR TITLE
Reactivate libdispatch tests on Linux

### PR DESCRIPTION
**Pull request by yln**
The swift-corelibs-libdispatch project had its build directory layout
changed [1] which silently deactivated the libdispatch tests on Linux,
because we use hard-coded paths to look for the required libdispatch
artifacts.  This change adapts the paths to the new layout.

Thanks to Jordan Rose for noticing and diagnosing this [2].
SR-11568

[2] https://bugs.swift.org/browse/SR-11568
[1] https://github.com/apple/swift-corelibs-libdispatch/pull/515

### CODE REVIEW

**Comment by yln** [comment]
@swift-ci Please test

**Comment by yln** [comment]
```
build/linux/libdispatch-linux-x86_64/src/libdispatch.so ->
build/linux/libdispatch-linux-x86_64/libdispatch.so

build/linux/libdispatch-linux-x86_64/src/libswiftDispatch.so ->
build/linux/libdispatch-linux-x86_64/libswiftDispatch.so

build/linux/libdispatch-linux-x86_64/src/swift/Dispatch.swiftmodule ->
build/linux/libdispatch-linux-x86_64/src/swift/swift/Dispatch.swiftmodule


➤ find build/linux/ -name libdispatch.so
build/linux/libdispatch-linux-x86_64/libdispatch.so
build/linux/swift-linux-x86_64/libdispatch-prefix/lib/libdispatch.so
build/linux/swift-linux-x86_64/libdispatch-prefix/src/libdispatch-build/libdispatch.so

➤ find build/linux/ -name libswiftDispatch.so
build/linux/libdispatch-linux-x86_64/libswiftDispatch.so

➤ find build/linux/ -name Dispatch.swiftmodule
build/linux/libdispatch-linux-x86_64/src/swift/swift/Dispatch.swiftmodule
build/linux/libdispatch_static-linux-x86_64/src/swift/swift/Dispatch.swiftmodule
```

**Comment by yln** [comment]
@swift-ci Please test

**Comment by yln** [comment]
lit now reports the libdispatch feature again.
```
lit.py: /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/swift/test/lit.cfg:1678: note:
Running tests on Ubuntu-16.04
lit.py: /home/buildnode/jenkins/workspace/swift-PR-Linux/branch-master/swift/test/lit.cfg:1688: note:
Available features: <more text>, executable_test, fuzzer_runtime, **libdispatch**, lldb, <more text>
```

Among others the following test is now executed again on Linux.
```
PASS: Swift(linux-x86_64) :: stdlib/DispatchData.swift (7550 of 12291)
```

Please merge (despite other unrelated failures).

**Comment by yln** [comment]
@shahmishal It would be a nice feature of CI to get a warning in the form of a PR comment if the number of executed tests decreases for a given PR.  This would help prevent "accidental disabling" of test.

**Comment by gottesmm** [comment]
Hey J!

Is there any way we could make it so that if this happens again, we get an error instead of silently not running the tests?

**Comment by yln** [comment]
https://github.com/apple/swift-corelibs-libdispatch/pull/515
https://bugs.swift.org/browse/SR-11568

**Comment by jrose-apple** [comment]
Ah, no credit to me. I just tagged you when that bug was filed.

File: test/lit.cfg [hunk 1]
```diff
@@ -1020,16 +1020,16 @@ elif (run_os in ['linux-gnu', 'linux-gnueabihf', 'freebsd', 'windows-cygnus', 'w
     config.target_runtime = "native"
     config.target_swift_autolink_extract = inferSwiftBinary("swift-autolink-extract")
 
-    libdispatch_artifact_dir = make_path(config.libdispatch_build_path, 'src')
+    libdispatch_artifact_dir = config.libdispatch_build_path
+    libdispatch_swift_module_dir = make_path(libdispatch_artifact_dir, 'src', 'swift', 'swift')
```
Comment (jrose-apple) [inline test/lit.cfg]: This path seems…worse? Why the double nesting?

Comment (yln) [reply test/lit.cfg]: I have no idea.  I just built under docker/Linux and looked where things ended up.
See my comment above: https://github.com/apple/swift/pull/27940#issuecomment-547635227

**Comment by yln** [comment]
> Hey J! Is there any way we could make it so that if this happens again, we get an error instead of silently not running the tests?

I can certainly come up with something, but I believe that this is general issue (not just libdispatch), best dealt with at the CI level.

File: test/lit.cfg (hunk 1 above)
Comment (jrose-apple) [reply test/lit.cfg]: I think it's worth fixing the path where things get built rather than adjusting the test logic to look there.

**Comment by gottesmm** [comment]
What about if we add an environment variable that the CI sets that would cause us to error if we don't find it. (I.e. SWIFT_CI=1).

File: test/lit.cfg (hunk 1 above)
Comment (compnerd) [reply test/lit.cfg]: The "double" is because the swift overlay is in src/swift, and the swift modules/docs/interfaces are emitted into a swift directory.

**Comment by buttaface** [comment]
I found and reported this issue, as I had periodically been running the tests locally since the beginning of the year and noticed these disabled tests when looking at prior test logs. I could have easily submitted this patch fix myself, but it wouldn't have been robust to any further moves in the libdispatch build directory. Since I got no response when I mentioned the regression on that libdispatch pull, I filed the bug so that the devs involved would actually discuss:

1. where the libdispatch build artifacts should go, and if that's final or might change again
2. whether lit.cfg should simply disable libdispatch testing alone if those libraries aren't found again, as done now and missed, or error and stop the whole test run

Without nailing those two details down, this is going to keep silently breaking, as mentioned by @gottesmm.

**Comment by buttaface** [comment]
Sounds good, that should fix it.

**Comment by jrose-apple** [comment]
Swift may not care where the products go, but corelibs-libdispatch is still putting them in a silly place.

**Comment by compnerd** [comment]
I'm happy to change them in a follow up change to use a different path for the modules.  Right now, we have multiple layers of things working around limitations of previous approaches, and keeping everything working is difficult enough.  I'm pretty close to migrating Foundation, which should help remove some of this technical debt in the setup and making it easier to relocate the swift modules.

**Comment by yln** [comment]
@swift-ci Please test macOS

**Comment by yln** [comment]
> What about if we add an environment variable that the CI sets that would cause us to error if we don't find it. (I.e. SWIFT_CI=1).

@gottesmm 
I added this.  We need to ensure this environment variable is added on our bots.

> I don't think that swift really needs to care about where the products go. Adjusting the path here to enable the tests first is a good approach. Subsequently, we should change the lit configuration to be substituted by CMake via `lit.cfg.in` to use the generator expressions to determine the location of the libdispatch content. We now expose them through export targets, and we can do something like `$<TARGET_FILE_DIR:swiftDispatch>` to get the location of the `.dll`/`.so`/`.dylib` and `$<INTERFACE_INCLUDE_DIRECTORIES:swiftDispatch>` to get the location of the swift module.

@compnerd 
I would need further guidance on how to do this.

**Comment by yln** [comment]
@swift-ci Please test

**Comment by gottesmm** [comment]
@yln instead of using an environment variable or the like, I think @shahmishal would prefer to have some sort of variable set in lit.

Can you also change the presets to set this variable as well?

**Comment by yln** [comment]
@swift-ci Please test

**Comment by yln** [comment]
@swift-ci Please test Linux

**Comment by yln** [comment]
`--lit-args=...` specified on the command line or presets are not respected:
https://bugs.swift.org/browse/SR-11690

I removed my last 2 commits.  I think wiring through CMake is the right solution.  Once we have this maybe we can make building libdispatch mandatory on platforms that require it?

The swift_ci env var/lit param seems very brittle.  It is also a specific solution for a general problem.  I think this should be handled at the CI/[Jenkins](https://plugins.jenkins.io/test-results-analyzer) level outside of the messiness of build-script.

**Comment by yln** [comment]
*ping*

I would like to merge this so we start executing these tests on Linux again.

The two attempted, specific fixes have their own set of issues.  My current understanding is that all `--lit-args` specified in presets are ignored. 
https://bugs.swift.org/browse/SR-11690

My hope is that "wiring through CMake" as compnerd suggested will lower the chance of silently disabling the libdispatch tests.

I also want to re-iterate the my opinion that the "silent disabling of tests" is a general problem; it would be good to have a general solution:
https://github.com/apple/swift/pull/27940#issuecomment-547685009

**Review by finagolfin — APPROVED** [review]
Confirmed that this pull enables the libdispatch and tsan tests again by building with the 11/08 snapshot on Android. I agree that the larger issues can be fixed later.

**Comment by yln** [comment]
Thanks @buttaface. And also thanks for bringing our attention to this in the first place.

**Comment by buttaface** [comment]
@yln, I think this broke a couple months ago again, likely because #38888 split up the build pipeline so that cmark/llvm/swift are built and tested first.

The way it worked before was that all the `build-script-impl` products up through llbuild were built first before any tests were run, but with this new split, the compiler validation suite no longer finds a built libdispatch and the libdispatch and tsan tests from the compiler validation suite are not run on the linux CI.

**Comment by yln** [comment]
Fixed by @edymtt in https://github.com/apple/swift/pull/41135

**Pull request merged**